feat(autoware_sensing_msgs): add PointCloud meta info msgs

amadeuszsz opened the pull request:
## Description

Add a new message for storing meta information of concatenated point cloud.

Required for https://github.com/autowarefoundation/autoware_universe/pull/10851.

## Relevant Links

 - https://github.com/orgs/autowarefoundation/discussions/6022

## How was this PR tested?

## Notes for reviewers

None.

## Effects on system behavior

None.

vividf commented on autoware_sensing_msgs/msg/ConcatenatedPointCloudInfo.msg:
```diff
@@ -0,0 +1,21 @@
+uint32 row_step                     # Length of a row in bytes
+
+bool is_dense                       # True if there are no invalid points
+
```
I would like to add 5 more values.
bool concatenation_success
double concatenated_pointcloud_timestamp
bool use_advaced_mode 
double min_reference_timestamp
double max_reference_timestamp

amadeuszsz replied:
Thanks for you input! Addressed in https://github.com/autowarefoundation/autoware_msgs/pull/142/commits/94270162cb7e24b788b7ced9417c5f4aca6c025a according to our discussion.

vividf reviewed (approved):
LGTM

vividf commented:
@mitsudome-r 
Some of the fields will be used in the `fusion_node` in `autoware_image_projection_based_fusion` (perception) and could be useful for other downstream module that needs concatenation information.

mitsudome-r commented:
Since this is going to be a change for component interface, we should also update the documentation page:
 - https://autowarefoundation.github.io/autoware-documentation/main/design/autoware-architecture/sensing/data-types/point-cloud/
 - https://autowarefoundation.github.io/autoware-documentation/main/design/autoware-interfaces/components/sensing/

@amadeuszsz Could you create PRs for the documentation as well

mitsudome-r commented:
> Extra feature (sensor ID) for ML algorithms.
Could you elaborate on this? Do you mean we want to have sensor ID encoded as a input to ML detection modules, or is it required for some kind of preprocessing?

yukkysaito commented:
@amadeuszsz I have a few questions:

- Is it mandatory to publish this as an official output interface of the Sensing Component?
- The individual LiDAR topics as well as the concatenated pointcloud topic are already defined as interfaces. Wouldn’t it be possible to reconstruct the necessary information from these existing topics? (If the LiDAR ID is needed, wouldn’t subscribing to the individual topics be sufficient?)

Please understand that changing the official output interface of a component affects the overall architecture of Autoware. As such, it will require time for discussion and review.

amadeuszsz commented:
@mitsudome-r 
> > Extra feature (sensor ID) for ML algorithms.
> 
> Could you elaborate on this? Do you mean we want to have sensor ID encoded as a input to ML detection modules, or is it required for some kind of preprocessing?

In ML modules, e.g. [CenterPoint's encoder](https://github.com/autowarefoundation/autoware_universe/blob/d67fe97ce62ff73eb8660471e9aa1a876d505921/perception/autoware_lidar_centerpoint/lib/preprocess/preprocess_kernel.cu#L238), might potentially use this input as an extra feature alongside other features.

> Since this is going to be a change for component interface, we should also update the documentation page:
> 
> * https://autowarefoundation.github.io/autoware-documentation/main/design/autoware-architecture/sensing/data-types/point-cloud/
> * https://autowarefoundation.github.io/autoware-documentation/main/design/autoware-interfaces/components/sensing/
> 
> @amadeuszsz Could you create PRs for the documentation as well?

Yes, after resolving all the review's comments, I will address the documentation.

@yukkysaito 
> * Is it mandatory to publish this as an official output interface of the Sensing Component?

Because `*_internal_msgs` supposed to be used only within desired components, `autoware_sensing_msgs` seems appropriate placement. Additionally, this message belongs to sensing pipeline, but it is planned to be used in perception component. I can imagine that in future we might use this meta information also in sensing filters or localization.

> * The individual LiDAR topics as well as the concatenated pointcloud topic are already defined as interfaces. Wouldn’t it be possible to reconstruct the necessary information from these existing topics? (If the LiDAR ID is needed, wouldn’t subscribing to the individual topics be sufficient?)

In that case, concatenation process would need to be performed once again in desired perception algorithm.

> Please understand that changing the official output interface of a component affects the overall architecture of Autoware. As such, it will require time for discussion and review.

Thanks for pointing this! I assumed this extra topic as a non-invasive and lightweight solution. Adding this message will not break anything and can open doors for future development. Another approach, e.g. adding `sensor_id` to point types will break API and is memory non-effective. 

Apart from ML algorithms, there is also another points where this message finds utilization:
* Fusion pipeline. Initially we used diagnostics for reporting concatenation status, but subscription of `/diagnostics` is heavy. Using namespace for diagnostics looks like a trick, also I think Diagnostics msgs should be used only for diagnostic aggregation, not processing in components.
* Dataset pipeline. In order to generate same data as it is during the runtime, we want to integrate Autoware sensing pipeline into dataset creation process. We could also copy and modify sensing pipeline to generate such a meta information, but it is not effective from maintenance point of view. About this meta information itself - we want to reconstruct the source point clouds from annotated data. Why not to use single clouds instead? Labeling N clouds instead of a single concatenated cloud is not cost-effective.

amadeuszsz commented on autoware_sensing_msgs/msg/SourcePointCloudInfo.msg:
```diff
@@ -0,0 +1,27 @@
+uint32 height
+uint32 width
+
+# Describes the channels and their layout in the binary data blob.
+sensor_msgs/PointField[] fields
+
+bool is_bigendian      # Is this data bigendian?
+uint32 point_step      # Length of a point in bytes
+uint32 row_step        # Length of a row in bytes
```
For now, we don't need it. We placed this information because It is lightweight and let us handle meta information independently without PointCloud2.

amadeuszsz replied:
We don't need them. Also resultant point cloud with various fields configuration is invalid.
Addressed in https://github.com/autowarefoundation/autoware_msgs/pull/142/commits/97571a4660f237c2018ab39af30f900fd92c7760

amadeuszsz commented on autoware_sensing_msgs/msg/ConcatenatedPointCloudInfo.msg:
```diff
@@ -0,0 +1,26 @@
+uint32 width
+
+# Describes the channels and their layout in the binary data blob.
+sensor_msgs/PointField[] fields
```
In this case I would like to keep these fields. It is not heavy and let us handle meta information independently from PointCloud2 msg (all the information in one place).

amadeuszsz replied:
@mitsudome-r 
Agree, addressed in https://github.com/autowarefoundation/autoware_msgs/pull/142/commits/437db015b8152081dea2acbec998f5a95d499ebe

xmfcx replied:
This point is addressed in the latest revision 👍

amadeuszsz commented on autoware_sensing_msgs/msg/ConcatenatedPointCloudInfo.msg:
```diff
@@ -0,0 +1,26 @@
+# point clouds, including the overall point cloud structure and metadata for each source segment.
+
+# Time of sensor data acquisition, and the coordinate frame ID (for 3d points).
+std_msgs/Header header
```
Right, sorry for missing this description! It matches the child msg `SourcePointCloudInfo`, but not the concatenated one. Addressed in https://github.com/autowarefoundation/autoware_msgs/pull/142/commits/8bde8c86b86808db5d7bed1af4c23948292a0f5c

xmfcx replied:
This is addressed in the latest revision 👍

yukkysaito commented:
> I can imagine that in future we might use this meta information also in sensing filters or localization.

Do we have any specific plans or use cases in mind for how this meta information would be used in sensing filters or localization?

> In that case, concatenation process would need to be performed once again in desired perception algorithm.

If we use this in ML models like CenterPoint, wouldn’t we end up copying and reshaping the data anyway? In that sense, the overhead might be similar.

> Thanks for pointing this! I assumed this extra topic as a non-invasive and lightweight solution. Adding this message will not break anything and can open doors for future development. Another approach, e.g. adding sensor_id to point types will break API and is memory non-effective.

While I understand and appreciate the intent of this proposal, I also see several concerns that should be taken into account:
- Use of Autoware-specific message types
  - Replacing the standard ROS message type PointCloud with a custom Autoware message may reduce compatibility and limit adoption by the broader ROS community and users.
- Impact on Component Interface Compatibility
  - Autoware is designed with well-defined component interfaces, enabling flexible substitution and modularity. Modifying these interfaces may break compatibility with existing sensing components and undermine this adaptive architecture.
- Redundancy with Existing Sensing Components
  - Some of the proposed information already exists in current interfaces, which may introduce redundancy. If we go with a redundant design, we must at least ensure consistency and prevent conflicting information.

Architecture design often involves unavoidable trade-offs, and it’s rare to find a solution that satisfies all constraints perfectly. I believe it would be helpful for us to discuss and align on a practical compromise that balances the pros and cons of this change.

amadeuszsz commented:
@yukkysaito 

> > I can imagine that in future we might use this meta information also in sensing filters or localization.
> 
> Do we have any specific plans or use cases in mind for how this meta information would be used in sensing filters or localization?

There is no plans for now apart from perception component. I wanted to point that this message can be utilized in other components, therefore `*_internal_msgs` might be not the best destination.

> > In that case, concatenation process would need to be performed once again in desired perception algorithm.
> 
> If we use this in ML models like CenterPoint, wouldn’t we end up copying and reshaping the data anyway? In that sense, the overhead might be similar.

Not exactly. Having such a meta information allows us to inject sensor ID directly into model's inference pre-processing, e.g. [here](https://github.com/autowarefoundation/autoware_universe/blob/d67fe97ce62ff73eb8660471e9aa1a876d505921/perception/autoware_lidar_centerpoint/lib/preprocess/preprocess_kernel.cu#L318). I don't see more efficient way. Subscription all the point clouds separately in CenterPoint and perform concatenation over there does not seem attractive, especially due to complex matching strategy already performed in sensing's point cloud pre-processing pipeline.

> > Thanks for pointing this! I assumed this extra topic as a non-invasive and lightweight solution. Adding this message will not break anything and can open doors for future development. Another approach, e.g. adding sensor_id to point types will break API and is memory non-effective.
> 
> While I understand and appreciate the intent of this proposal, I also see several concerns that should be taken into account:
> 
> * Use of Autoware-specific message types
>   
>   * Replacing the standard ROS message type PointCloud with a custom Autoware message may reduce compatibility and limit adoption by the broader ROS community and users.
> * Impact on Component Interface Compatibility
>   
>   * Autoware is designed with well-defined component interfaces, enabling flexible substitution and modularity. Modifying these interfaces may break compatibility with existing sensing components and undermine this adaptive architecture.
> * Redundancy with Existing Sensing Components
>   
>   * Some of the proposed information already exists in current interfaces, which may introduce redundancy. If we go with a redundant design, we must at least ensure consistency and prevent conflicting information.
> 
> Architecture design often involves unavoidable trade-offs, and it’s rare to find a solution that satisfies all constraints perfectly. I believe it would be helpful for us to discuss and align on a practical compromise that balances the pros and cons of this change.

I think we might misunderstood each other. This message does not replace PointCloud2 message. It does not contain `data` field. There is no Point Cloud. It acts almost like a `sensing_msgs/CameraInfo` msg. It is just a message generated **extra** in concatenation process and published **separately**. Some fields overlaps with PointCloud2 because it does not cost us much and allows to store all the information about point cloud in one place.

amadeuszsz commented on autoware_sensing_msgs/msg/ConcatenatedPointCloudInfo.msg:
```diff
@@ -0,0 +1,26 @@
+bool is_dense                                    # True if there are no invalid points
+
+bool concatenation_success                       # Whether concatenation was successful
+bool used_advanced_mode                          # Whether used advanced mode
```
Since we might have other strategies in the future, the configuration might vary. For this reason we would prefer to store configuration as a raw data, and later perform deserialization.
Addressed in https://github.com/autowarefoundation/autoware_msgs/pull/142/commits/189180854633d7fd29ce9b936febf196f906302c
This design prevents radical msg updates in the future, just adding `uint8 STRATEGY_* = N` will be enough.

amadeuszsz commented:
@mitsudome-r 
Thank you for your review. Already addressed all the comments.

Regarding

> Since this is going to be a change for component interface, we should also update the documentation page:
> 
> * https://autowarefoundation.github.io/autoware-documentation/main/design/autoware-architecture/sensing/data-types/point-cloud/
> * https://autowarefoundation.github.io/autoware-documentation/main/design/autoware-interfaces/components/sensing/
> 
> @amadeuszsz Could you create PRs for the documentation as well?

and my previous answer (that I will update docs as well), I would also clarify if we understood each other correctly. This message does not replace PointCloud2, it is an extra message. I can update documentation with information about this extra output topic, but I'm not sure if your intention is the same after this clarification.

Also if PR will be approved, please give me a bit time as I would like to test universe node with approved design 🙏🏻

drwnz commented:
@yukkysaito @mitsudome-r 
Thank you for reviewing this proposal.

This proposal does not affect current interface or data pipelines in anyway, and as such is a non-breaking change, The background is that we recognized the current **concatenated** poitntcloud, using standard PointCloud2 message type and Autoware PointField definitions, misses some key information about the constituent data (namely, the source individual source sensors and the type of concatenation used).

This isn't just a "nice to have" feature, but a need identified from perception requirement to adopt new perception models, and more importantly, for creating ML pipelines for datasets using Autoware. This isn't the first time I have worked on this problem - previously, I addressed it by entirely re-creating the Autoware modules to process data in dataset creation.

I will set a meeting with you to explain in more detail the background, as it touches on some internal TIER IV projects which may not be appropriate to discuss here.

drwnz reviewed (approved):
LGTM!

xmfcx commented on autoware_sensing_msgs/msg/ConcatenatedPointCloudInfo.msg:
```diff
@@ -0,0 +1,17 @@
+# point clouds, including the overall point cloud structure and metadata for each source segment.
+
+# Available matching strategies for source point clouds concatenation
+uint8 STRATEGY_NAIVE = 0     # Assigns the oldest stamp of the sensor that was concatenated
+uint8 STRATEGY_ADVANCED = 1  # Assigns the oldest stamp of the sensor that was concatenated
```
Could you add some information on how these strategies work/differ? Maybe in the readme file?

xmfcx replied:
I prefer **A** (update [there](https://autowarefoundation.github.io/autoware_universe/main/sensing/autoware_pointcloud_preprocessor/docs/concatenate-data/#concatenation-strategies) and link from the readme here) with very simple descriptions of strategies in the readme of this package too.

amadeuszsz replied:
Addressed in https://github.com/autowarefoundation/autoware_universe/pull/10994 and https://github.com/autowarefoundation/autoware_msgs/pull/142/commits/f9d93b89ea9361509367ad4aeba94247c4f248f7

xmfcx commented on autoware_sensing_msgs/msg/ConcatenatedPointCloudInfo.msg:
```diff
@@ -0,0 +1,17 @@
+# The concatenated point cloud header, where timestamp assignment depends on matching strategy
+std_msgs/Header header
+
+bool concatenation_success                       # Whether concatenation was successful
```
Is this field for, if some point clouds are missing from the concat point cloud? Or are there other cases that could trigger this condition?

amadeuszsz replied:
Depends on strategy. Strategies may have configuration which describe requirements for concatenation, e.g. maximum latency. I see it this way:
- Strict strategy set, concatenated point cloud misses one point cloud, `concatenation_success = false`.
- Loose strategy set, concatenated point cloud misses one point cloud, `concatenation_success = true`.

xmfcx replied:
Then maybe it could make sense to include this info in the message directly.
For this, we can add the following field:
```
string[] missing_cloud_topics               # List of the point cloud topics missing from the concatenated cloud
```

And the concatenation_success can be evaluated with `if len(missing_cloud_topics) > 0`

amadeuszsz replied:
Not a big fan as `idx_end` says to me that point cloud is missing. `cloud_is_missing` seems redundant (subjective though 😄). Do we have a scenario where `idx_end = 0` and `cloud_is_missing = false`?

Some other idea came to my mind:
We can remove `idx_end` and have `size`. This will get rid of `idx_end` inclusive/exclusive considerations as users usually know how to operate with `size`-named variables (`<` operator instead of `<=` operator).
Also if not using `cloud_is_missing`, `size` seems more intuitive than `idx_end` for interfering if cloud exist or not. On the top of that, there is no potential issue where `idx_begin` > `idx_end`. Name `size` can be exchanged with `length` as `size` might suggest byte-size.

After my considerations, which idea do you prefer? Current, your or proposed?

xmfcx replied:
I think **option C** is alright. As long as the cloud is OK, the end point can check the size and decide to handle "empty point cloud" case.

And I agree, `STATUS_MISSING` and `STATUS_TIMEOUT` have too much overlap for the concat node. So just keeping TIMEOUT should suffice.

xmfcx commented on autoware_sensing_msgs/msg/SourcePointCloudInfo.msg:
```diff
@@ -0,0 +1,13 @@
+# The starting index (inclusive) of the point cloud segment within the concatenated point cloud
+uint32 idx_begin
+
+# The ending index (exclusive) of the point cloud segment within the concatenated point cloud
+uint32 idx_end
```
I'm a little bit concerned on how this could be interpreted.

`idx_end` should mean the index of the ending point so it should be inclusive.

What are your thoughts on this? Are there prominent examples where this convention is followed?

xmfcx commented on autoware_sensing_msgs/msg/SourcePointCloudInfo.msg:
```diff
@@ -0,0 +1,20 @@
+# The starting index (inclusive) of the point cloud segment within the concatenated point cloud
+uint32 idx_begin
+
+# The lenght of the point cloud segment in the concatenated point cloud
```
```suggestion
# The length of the point cloud segment in the concatenated point cloud
```

xmfcx commented on autoware_sensing_msgs/README.md:
```diff
@@ -223,8 +221,47 @@ In the message file, it is a `float64[36]` array. We fill the indices at `i*7, i
+- `header`: Standard ROS header with timestamp and frame information for the concatenated point cloud
+- `concatenation_success`: Boolean indicating whether the concatenation process completed successfully
+- `matching_strategy`: Strategy used for concatenation (see constants below)
+- `matching_strategy_config`: Raw encoded configuration data specific to the matching strategy (decoded by the node)
```
Could you elaborate on `matching_strategy_config` as well? I couldn't find information [within the documentation page](https://autowarefoundation.github.io/autoware_universe/main/sensing/autoware_pointcloud_preprocessor/docs/concatenate-data/#concatenation-strategies).

amadeuszsz replied:
This will be added alongside with https://github.com/autowarefoundation/autoware_universe/pull/10851
Maybe can we add here a placeholder link (https://autowarefoundation.github.io/autoware_universe/main/sensing/autoware_pointcloud_preprocessor/docs/concatenate-data/#meta-information-topic) for `Meta information topic` section?

xmfcx replied:
I see, thanks, that should suffice 👍
